File: src/test/java/com/koddy/server/coffeechat/application/usecase/HandlePendingCoffeeChatUseCaseTest.java
package com.koddy.server.coffeechat.application.usecase;

import com.koddy.server.coffeechat.application.usecase.command.FinallyApprovePendingCoffeeChatCommand;
import com.koddy.server.coffeechat.application.usecase.command.FinallyCancelPendingCoffeeChatCommand;
import com.koddy.server.coffeechat.domain.event.BothNotification;
import com.koddy.server.coffeechat.domain.event.MenteeNotification;
import com.koddy.server.coffeechat.domain.model.CoffeeChat;
import com.koddy.server.coffeechat.domain.model.Strategy;
import com.koddy.server.coffeechat.domain.repository.CoffeeChatRepository;
import com.koddy.server.coffeechat.domain.service.CoffeeChatNotificationEventPublisher;
import com.koddy.server.common.UnitTest;
import com.koddy.server.common.fixture.CoffeeChatFixture;
import com.koddy.server.common.mock.fake.FakeEncryptor;
import com.koddy.server.global.utils.encrypt.Encryptor;
import com.koddy.server.member.domain.model.mentee.Mentee;
import com.koddy.server.member.domain.model.mentor.Mentor;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.context.ApplicationEventPublisher;

import java.time.LocalDateTime;

import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_FINALLY_APPROVE;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_FINALLY_CANCEL;
import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_1;
import static com.koddy.server.common.fixture.MentorFixture.MENTOR_1;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@DisplayName("CoffeeChat -> HandlePendingCoffeeChatUseCase 테스트")
class HandlePendingCoffeeChatUseCaseTest extends UnitTest {
    private final CoffeeChatRepository coffeeChatRepository = mock(CoffeeChatRepository.class);
    private final Encryptor encryptor = new FakeEncryptor();
    private final ApplicationEventPublisher applicationEventPublisher = mock(ApplicationEventPublisher.class);
    private final CoffeeChatNotificationEventPublisher eventPublisher = new CoffeeChatNotificationEventPublisher(applicationEventPublisher);
    private final HandlePendingCoffeeChatUseCase sut = new HandlePendingCoffeeChatUseCase(
            coffeeChatRepository,
            encryptor,
            eventPublisher
    );

    private final Mentee mentee = MENTEE_1.toDomain().apply(1L);
    private final Mentor mentor = MENTOR_1.toDomain().apply(2L);

    @Test
    @DisplayName("Pending 상태인 CoffeeChat에 대해서 멘토는 최종 취소한다")
    void finallyCancel() {
        // given
        final LocalDateTime start = LocalDateTime.of(2024, 2, 1, 9, 0);
        final CoffeeChat coffeeChat = CoffeeChatFixture.MentorFlow.suggestAndPending(start, start.plusMinutes(30), mentor, mentee).apply(1L);

        final FinallyCancelPendingCoffeeChatCommand command = new FinallyCancelPendingCoffeeChatCommand(mentor.getId(), coffeeChat.getId(), "거절...");
        given(coffeeChatRepository.getByIdAndMentorId(command.coffeeChatId(), command.mentorId())).willReturn(coffeeChat);

        // when
        sut.finallyCancel(command);

        // then
        assertAll(
                () -> verify(coffeeChatRepository, times(1)).getByIdAndMentorId(command.coffeeChatId(), command.mentorId()),
                () -> verify(applicationEventPublisher, times(1)).publishEvent(any(MenteeNotification.MentorFinallyCanceledFromMentorFlowEvent.class)),
                () -> assertThat(coffeeChat.getMentorId()).isEqualTo(mentor.getId()),
                () -> assertThat(coffeeChat.getMenteeId()).isEqualTo(mentee.getId()),
                () -> assertThat(coffeeChat.getStatus()).isEqualTo(MENTOR_FINALLY_CANCEL),
                () -> assertThat(coffeeChat.getReason().getApplyReason()).isNull(),
                () -> assertThat(coffeeChat.getReason().getSuggestReason()).isNotNull(),
                () -> assertThat(coffeeChat.getReason().getCancelReason()).isNotNull(),
                () -> assertThat(coffeeChat.getReason().getRejectReason()).isNull(),
                () -> assertThat(coffeeChat.getQuestion()).isNotNull(),
                () -> assertThat(coffeeChat.getReservation().getStart()).isEqualTo(start),
                () -> assertThat(coffeeChat.getReservation().getEnd()).isEqualTo(start.plusMinutes(30)),
                () -> assertThat(coffeeChat.getStrategy()).isNull()
        );
    }

    @Test
    @DisplayName("Pending 상태인 CoffeeChat에 대해서 멘토는 최종 수락한다")
    void finallyApprove() {
        // given
        final LocalDateTime start = LocalDateTime.of(2024, 2, 1, 9, 0);
        final CoffeeChat coffeeChat = CoffeeChatFixture.MentorFlow.suggestAndPending(start, start.plusMinutes(30), mentor, mentee).apply(1L);

        final FinallyApprovePendingCoffeeChatCommand command = new FinallyApprovePendingCoffeeChatCommand(
                mentor.getId(),
                coffeeChat.getId(),
                Strategy.Type.KAKAO_ID,
                "sjiwon"
        );
        given(coffeeChatRepository.getByIdAndMentorId(command.coffeeChatId(), command.mentorId())).willReturn(coffeeChat);

        // when
        sut.finallyApprove(command);

        // then
        assertAll(
                () -> verify(coffeeChatRepository, times(1)).getByIdAndMentorId(command.coffeeChatId(), command.mentorId()),
                () -> verify(applicationEventPublisher, times(1)).publishEvent(any(BothNotification.FinallyApprovedFromMentorFlowEvent.class)),
                () -> assertThat(coffeeChat.getMentorId()).isEqualTo(mentor.getId()),
                () -> assertThat(coffeeChat.getMenteeId()).isEqualTo(mentee.getId()),
                () -> assertThat(coffeeChat.getStatus()).isEqualTo(MENTOR_FINALLY_APPROVE),
                () -> assertThat(coffeeChat.getReason().getApplyReason()).isNull(),
                () -> assertThat(coffeeChat.getReason().getSuggestReason()).isNotNull(),
                () -> assertThat(coffeeChat.getReason().getCancelReason()).isNull(),
                () -> assertThat(coffeeChat.getReason().getRejectReason()).isNull(),
                () -> assertThat(coffeeChat.getQuestion()).isNotNull(),
                () -> assertThat(coffeeChat.getReservation().getStart()).isEqualTo(start),
                () -> assertThat(coffeeChat.getReservation().getEnd()).isEqualTo(start.plusMinutes(30)),
                () -> assertThat(coffeeChat.getStrategy().getType()).isEqualTo(command.type()),
                () -> assertThat(coffeeChat.getStrategy().getValue()).isNotEqualTo(command.value()),
                () -> assertThat(encryptor.decrypt(coffeeChat.getStrategy().getValue())).isEqualTo(command.value())
        );
    }
}

File: src/test/kotlin/com/koddy/server/coffeechat/application/usecase/HandlePendingCoffeeChatUseCaseTest.kt
package com.koddy.server.coffeechat.application.usecase

import com.koddy.server.coffeechat.application.usecase.command.FinallyApprovePendingCoffeeChatCommand
import com.koddy.server.coffeechat.application.usecase.command.FinallyCancelPendingCoffeeChatCommand
import com.koddy.server.coffeechat.domain.event.BothNotification
import com.koddy.server.coffeechat.domain.event.MenteeNotification
import com.koddy.server.coffeechat.domain.model.CoffeeChat
import com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_FINALLY_APPROVE
import com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_FINALLY_CANCEL
import com.koddy.server.coffeechat.domain.model.Strategy
import com.koddy.server.coffeechat.domain.service.CoffeeChatNotificationEventPublisher
import com.koddy.server.coffeechat.domain.service.CoffeeChatReader
import com.koddy.server.common.UnitTestKt
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_1주차_20_00_시작
import com.koddy.server.common.fixture.MenteeFixtureStore.menteeFixture
import com.koddy.server.common.fixture.MentorFixtureStore.mentorFixture
import com.koddy.server.common.fixture.MentorFlow
import com.koddy.server.common.mock.fake.FakeEncryptor
import com.koddy.server.member.domain.model.mentee.Mentee
import com.koddy.server.member.domain.model.mentor.Mentor
import io.kotest.assertions.assertSoftly
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.mockk.every
import io.mockk.justRun
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.springframework.context.ApplicationEventPublisher

@UnitTestKt
@DisplayName("CoffeeChat -> HandlePendingCoffeeChatUseCase 테스트")
internal class HandlePendingCoffeeChatUseCaseTest : DescribeSpec({
    val coffeeChatReader = mockk<CoffeeChatReader>()
    val encryptor = FakeEncryptor()
    val eventPublisher = mockk<ApplicationEventPublisher>()
    val coffeeChatNotificationEventPublisher = CoffeeChatNotificationEventPublisher(eventPublisher)
    val sut = HandlePendingCoffeeChatUseCase(
        coffeeChatReader,
        encryptor,
        coffeeChatNotificationEventPublisher,
    )

    val mentor: Mentor = mentorFixture(id = 1L).toDomain()
    val mentee: Mentee = menteeFixture(id = 2L).toDomain()

    describe("HandlePendingCoffeeChatUseCase's finallyCancel") {
        val coffeeChat: CoffeeChat = MentorFlow.suggestAndPending(id = 1L, fixture = 월요일_1주차_20_00_시작, mentor = mentor, mentee = mentee)

        context("멘티가 1차 수락한 커피챗에 대해서 멘토가 최종 취소하면") {
            val command = FinallyCancelPendingCoffeeChatCommand(
                mentorId = mentor.id,
                coffeeChatId = coffeeChat.id,
                cancelReason = "취소..",
            )
            every { coffeeChatReader.getByMentor(command.coffeeChatId, command.mentorId) } returns coffeeChat

            val slotEvent = slot<MenteeNotification>()
            justRun { eventPublisher.publishEvent(capture(slotEvent)) }

            it("커피챗은 최종 취소 상태가 되고 멘티에게 알림이 전송된다") {
                sut.finallyCancel(command)

                verify(exactly = 1) { coffeeChatReader.getByMentor(command.coffeeChatId, command.mentorId) }
                slotEvent.captured shouldBe MenteeNotification.MentorFinallyCanceledFromMentorFlowEvent(
                    menteeId = coffeeChat.menteeId,
                    coffeeChatId = coffeeChat.id,
                )
                assertSoftly(coffeeChat) {
                    // Effected
                    status shouldBe MENTOR_FINALLY_CANCEL
                    reason.cancelReason shouldBe command.cancelReason

                    // Not Effected
                    mentorId shouldBe mentor.id
                    menteeId shouldBe mentee.id
                    cancelBy shouldBe null
                    reason.applyReason shouldBe null
                    reason.suggestReason shouldNotBe null
                    reason.rejectReason shouldBe null
                    question shouldNotBe null
                    reservation!!.start shouldBe 월요일_1주차_20_00_시작.start
                    reservation!!.end shouldBe 월요일_1주차_20_00_시작.end
                    strategy shouldBe null
                }
            }
        }
    }

    describe("HandlePendingCoffeeChatUseCase's finallyApprove") {
        val coffeeChat: CoffeeChat = MentorFlow.suggestAndPending(id = 1L, fixture = 월요일_1주차_20_00_시작, mentor = mentor, mentee = mentee)

        context("멘티가 1차 수락한 커피챗에 대해서 멘토가 최종 수락하면") {
            val command = FinallyApprovePendingCoffeeChatCommand(
                mentorId = mentor.id,
                coffeeChatId = coffeeChat.id,
                type = Strategy.Type.KAKAO_ID,
                value = "sjiwon",
            )
            every { coffeeChatReader.getByMentor(command.coffeeChatId, command.mentorId) } returns coffeeChat

            val slotEvent = slot<BothNotification>()
            justRun { eventPublisher.publishEvent(capture(slotEvent)) }

            it("커피챗이 예정되고 멘토 & 멘티에게 알림이 전송된다") {
                sut.finallyApprove(command)

                verify(exactly = 1) { coffeeChatReader.getByMentor(command.coffeeChatId, command.mentorId) }
                slotEvent.captured shouldBe BothNotification.FinallyApprovedFromMentorFlowEvent(
                    menteeId = coffeeChat.menteeId,
                    mentorId = coffeeChat.mentorId,
                    coffeeChatId = coffeeChat.id,
                )
                assertSoftly(coffeeChat) {
                    // Effected
                    status shouldBe MENTOR_FINALLY_APPROVE
                    strategy!!.type shouldBe command.type
                    strategy!!.value shouldNotBe command.value
                    encryptor.decrypt(strategy!!.value) shouldBe command.value

                    // Not Effected
                    mentorId shouldBe mentor.id
                    menteeId shouldBe mentee.id
                    cancelBy shouldBe null
                    reason.applyReason shouldBe null
                    reason.suggestReason shouldNotBe null
                    reason.cancelReason shouldBe null
                    reason.rejectReason shouldBe null
                    question shouldNotBe null
                    reservation!!.start shouldBe 월요일_1주차_20_00_시작.start
                    reservation!!.end shouldBe 월요일_1주차_20_00_시작.end
                }
            }
        }
    }
})
